### PR TITLE
RFC- do not require FE review group manual approval for changes to /src/applications 

### PR DESCRIPTION
**Pull request by rianfowler**
## Overview 

- due to time constraints and lack of familiarity, FE review group is not creating useful feedback for application specific code 
- FE review group's time could be better spent focusing on the common components in vets-website like the shared css, build process, or site wide components 

## Review instructions 

- Add a comment providing pros, cons, and concerns with this approach. 
- Review the configuration carefully- [example](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-syntax)
- Add an approving review if approved. Add needs changes if you disapprove (everyone has a veto) 
- PR will remain open until 12/17/2019

I will add a documented decision [here](https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/tree/master/packages/documentation/src/pages/platform/front-end-standards/documented-decisions).

### CODE REVIEW

File: .github/CODEOWNERS
Comment (rianfowler) [inline .github/CODEOWNERS]: All files in the root directory excluding subdirectories

File: .github/CODEOWNERS
Comment (rianfowler) [inline .github/CODEOWNERS]: All files in the /src directory excluding subdirectories

**Review by rianfowler — COMMENTED** [review]
This should exclude /src/applications from requiring a review from frontend-review-group

**Comment by rianfowler** [comment]
Related: 

- We are evaluating third party code quality tools as part of [this](https://github.com/department-of-veterans-affairs/va.gov-team/issues/4091) epic
- Could delay this until automatic code quality enforcement is in place

**Comment by gunsch** [comment]
Pros:
* [med/large] clears up FE review time, better for FE team.
* [med/large] more autonomy for VFS teams, we stay out of their way where we're not needed

Cons:
* [tiny] CODEOWNERS config is awful
* [large?] if code quality declines, app becomes awful, app is eventually handed off, we might be left holding the bag (as platform). This is a VSP-wide issue we already experience a number of ways, review is currently _a_ mitigation towards it that I'd hate to lose.

  buuuuut, it's unclear to me how much of a mitigation that is in practice, and how much we could supplement that mitigation with better github/code-quality tools.

I'm tentatively supportive, but... have we run this by others? (DEPO / Brandon / Rachael). I can if not.

**Review by cvalarida — APPROVED** [review]
I'll echo everything @gunsch said, but I can also speak (type? :thinking:) to the code quality issue.

## On ensuring code quality
As it stands right now, we're honestly doing very little to ensure code quality during PR reviews. Because I don't know the broader architecture of many of the apps I'm reviewing, I have very little context on how any given PR might be improved aside from some pretty basic suggestions. ("Consider using a ternary here." "This might be cleaner with recursion." "What do you think about calling this `moreObviousVarName` instead?" That kind of thing.)

While those kinds of suggestions are useful periodically, I don't think their benefit outweighs the burden on the platform team.

Even if I _did_ have a better idea of how the architecture might be improved, figuring that out and communicating it takes a lot of time. (I've been guilty of doing this more than I'd like to admit :sweat_smile: )

To ensure a certain level of quality, I think we should investigate and leverage automated tools (like we've already talked about). These could include:
- Checking for code complexity / code smells (Code Climate)
- Checking for adequate documentation (JSDOC eslint plugin or something)
  - Not convinced how helpful this would be since I foresee just ensuring the docs are there doesn't help with keeping them up-to-date
- Static typing
  - This way we don't have to hunt down the shape of a variable and how it's been changed through a series of components or functions; we should know what we need to know (in most cases) from the type, if done well (which seems enforceable, but I don't actually know)
  - This is also a whole other can of worms we don't have to open up right now; I present it now only as something to consider
- Checking for certain paradigms
  - One of the concerns with static typing approach is the fundamental problem it solves (at least as I outlined above); namely, if we have to hunt down a  complex data structure through multiple components or functions, there's probably a better architecture we could employ
  - I have _no_ idea if this is enforceable at all! :sweat_smile: 

## Conclusion
I'm in favor of this. We definitely should make a list of the outcomes we believe to be important (bundle size, etc.) and come up with ways to automatically enforce _those_ though. Whether we do this before or after we lift the code ownership, I don't think I have an opinion.

## Caveat
One thing we should be aware of is that this also allows Team A to modify Team B's applications with only peer approval from Team A engineers. This _shouldn't_ be a problem, but we may consider a more nuanced approach down the road if this becomes an issue. (Discovering that it _is_ an issue might be tricky, though... :thinking: )

That said, I don't know who all's on what teams and what all their apps are, so...I don't think this is a new problem; I'm not personally likely to stop a PR from Team A that modifies Team B's stuff because I don't know:
- That this engineer is on Team A
- That the files in this PR actually affect Team B's work

**Comment by bkjohnson** [comment]
Besides the syntax change which Chris mentioned, the one change I might like to see is that we still get tagged for changes to any `*.css` or `*.scss` files in `src/applications` so that we can ensure that nobody is trying to circumvent the design system.

As for the caveat Chris mentioned - this could likely be avoided with the move to microfrontends down the road.


### Code Quality

Part of this is echoing Chris's thoughts above.

I would like the code quality to not drop (and ideally improve), but the circumstances of our review as gatekeeper limit our ability to do this somewhat.  Each team may come up with their own patterns, and those patterns may make sense in the context of the rest of the app's organization.  Since we don't have a lot of context behind things, I only feel comfortable critiquing relatively small pieces.  Having the authority to do larger periodic audits and fix things that we deem to be of poor quality might be helpful, but that would also be a bit of a burden.

**Comment by gunsch** [comment]
ooooh, big +1 to @bkjohnson 's point about CSS --- not allowing CSS by default (or if we can force CSS files to require FE review) is a great idea. if we believe our components and default CSS should be sufficient, then let's lock that down too

**Comment by jbalboni** [comment]
I just wanted to add that the concern @cvalarida raised about other teams modifying app code they don't own is something that's important to me as an application developer. I really want to know if anyone (including VSP folks!) modifies the VAOS application folder. And the temptation to make changes in apps you don't own is definitely there; we almost made a change to a component in the facility locator app the other day. We thought better of it and asked, but nobody is perfect under deadline crunches and other pressures that app teams can face.

**Comment by rianfowler** [comment]
@gunsch 

> have we run this by other leads? (DEPO / Brandon / Rachael). I can if not.

Brandon is aware but we haven't run this by DEPO or Rachael yet. We aren't reviewing for maintainability right now so changing this won't have an impact to that. 

I think "maintainability" is a meaningless term and we should delegate code quality enforcement to a third party tool: https://github.com/department-of-veterans-affairs/va.gov-team/issues/4091

**Comment by rianfowler** [comment]
@cvalarida 

> Checking for adequate documentation (JSDOC eslint plugin or something)

Can you make a ticket for this and summarize what you think it would look like? 

> Static typing

I think static typing is up for debate. My rough plan was to bring it up once we made some progress on micro frontends / app isolation. We could potentially make the platform code static typed. I've never worked on a static typed FE app so I don't have many opinions. I think the core problem with code quality on vets-website stem from the complete lack of abstractions and I'm not sure using static typing will solve that on its own. 

I think the biggest reason not to use static typed js is that it shrinks the pool of developers who can work on the website significantly. 

> Checking for certain paradigms

Some work for this has been outlined in the backlog for a long time: 

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/418
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/235
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/233
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/233

> Caveat

- I think we'll modify this approach to ensure every part of the site has a code owner. We can assign Code Owners to apps and reviews can fall through to the FE review group when no group is available.

File: .github/CODEOWNERS
Comment (rianfowler) [reply .github/CODEOWNERS]: We can try it but there's no exclusion syntax documented [here]( https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-syntax) so I don't think it's the same as gitignore.

Code Owners is a github feature. gitignore is a git feature.

**Comment by rianfowler** [comment]
@bkjohnson @gunsch What are we enforcing when we review CSS? 

- There's no a performance concern- no app specific CSS goes into the global CSS.  
- Consistency should be enforced by the design / ux group. AFAIK each app goes through a design review and we probably wouldn't be able to enforce "design consistency" by ensuring they're using the design system. 
- GIBCT and facility locator don't use the design system (formation) and in most of the time we don't try to make them refactor all of the app's CSS to use it.

**Comment by bkjohnson** [comment]
When I review CSS I try to see if the design system could be used instead - my goal has been to try make sure teams use what is already available without introducing new custom CSS. This often means that I'm acting as a "documentation evangelist" by pointing devs to the design system when I see that they are duplicating something that it already covers.

Sometimes the design system is effective in this, but @jbalboni and I just had a conversation about a scenario where the design system is a bit too limiting and trying to force it onto existing code that doesn't use it can be a burden.  So with that in mind I haven't been incredibly adamant about using the design system all the time.

Also, it sounds like the design review you're talking about happens afterwards on a periodic basis where dozens of PRs may have gotten through which introduced lots of tightly coupled markup and custom CSS.  If the peer review process had some consideration to the design system I would be ok with that, but it often doesn't, which leaves us in the position where we can mitigate some pain of that design review/refactoring.

File: .github/CODEOWNERS
Comment (rianfowler) [inline .github/CODEOWNERS]: in CODEOWNERS, only one required approval is triggered on a PR at a time. Also, 

> the last matching pattern takes the most precedence

With this config, the FE review group is the default approval needed for any change to the repo.

Anything change the /app/vaos folder requires approval by the vaos group and not the FE review group.

Any change to scss or css require approval by the FE review group.

Comment (rianfowler) [reply .github/CODEOWNERS]: @jbalboni :  fyi- we're using your group as an example. 

@department-of-veterans-affairs/frontend-review-group @gunsch how do you think we should get the rest of the teams groups set up? looks like nancy set up the vaos group so maybe this is already in progress.

Comment (rianfowler) [reply .github/CODEOWNERS]: ah okay I found this: https://github.com/orgs/department-of-veterans-affairs/teams/vets-gov-write/teams

So I think we could ask teams to add "review groups" as suub groups to their contract groups. we should probably move the FE review group into this https://github.com/orgs/department-of-veterans-affairs/teams/vets-gov-write group

**Comment by cvalarida** [comment]
@bkjohnson Using the design system in apps solves two problems:
1. A consistent(ish) experience for Veterans (and other users of the site)
1. A smaller style file shipped to the browser by avoiding duplicate code

The first point should be solved by a solid design review, as @rianfowler mentioned. It's not really the responsibility of the platform to be a gatekeeper for that kind of thing. As it stands right now, it seems like that's been helpful, but I'll be honest: I don't want to review application styles.

For the second point, I think this is where we automatically enforce size restrictions. Something something performance budgets.

Also for the record, I've been dreaming about making a tool that analyzes CSS / SCSS files for duplicate rules and the like. So theoretically, we could feed in a style file and see that we could remove a certain rule and use these two or three from the design system in its place. I've tried to find a tool that does just this, but...I'm so bad at finding tools. :grimacing:

**Comment by rianfowler** [comment]
> I've been dreaming about making a tool that analyzes CSS / SCSS files for duplicate rules and the like. 

@cvalarida not sure how SCSS does it but I go to generating a AST when I see this.

https://github.com/shawnbot/sast

File: .github/CODEOWNERS
Comment (gunsch) [reply .github/CODEOWNERS]: I'd check with @nwalke about intent/administration of GH teams structure as that gets built out. I don't know enough about which GH teams we do/don't have already, particularly for GH teams that correspond to VFS teams.

**Comment by leahbannon** [comment]
This looks reasonable to me based on the trade-offs. Can y'all provide a couple plain language examples of the kinds of problems we're currently catching? This will help us socialize this with leadership. The CSS/design system example is a good one.

Also, these are some lingering questions folks flagged that I'm curious about:
- Write up your hypotheses/KPIs first — Good point from @cvalarida: "We definitely should make a list of the outcomes we believe to be important (bundle size, etc.) and come up with ways to automatically enforce those though." As part of that, from @bkjohnson "Plan to do larger periodic audits and fix things that we deem to be of poor quality might be helpful, but that would also be a bit of a burden." 
- CSS review concerns — Resolved through design reviews and maintaining control over global CSS (no app specific CSS goes into the global CSS). Need @KevinMHoffmanUSDS's thoughts on this. We may want to pull in someone from FE team for design reviews if that’s not happening already? To keep catching design system duplication, as part of the [automated reviews super epic](https://github.com/department-of-veterans-affairs/va.gov-team/issues/4091), are y'all planning to add an automated check like the example Rian provided with https://github.com/shawnbot/sast for this as well?
- Automated code reviews — It seems like there's consensus that it's not worth on this change before finishing the automated reviews epic (and that continuing to do FE reviews greatly hinders progress on this). Is that accurate? What would the rough timeline be like for building this out if the team isn't also doing FE reviews?
- Team-level control — Are the GitHub teams currently organized in a way that supports this? Should we/do we plan to restrict the people who can approve FE PRs for their team’s code to FE devs? 
- Size restrictions — For reviews that are currently enforcing “A smaller style file shipped to the browser by avoiding duplicate code” @cvalarida recommended automatically enforce size restrictions. Was that suggestion incorporated? It sounds like a good one to me.

**Comment by KevinMHoffmanUSDS** [comment]
> CSS review concerns — Resolved through design reviews and maintaining control over global CSS (no app specific CSS goes into the global CSS). Need @KevinMHoffmanUSDS's thoughts on this. We may want to pull in someone from FE team for design reviews if that’s not happening already?

@leahbannon My thoughts are as follows: 
- Many CSS consistency issues (using the existing system) can be caught before CSS is written within the proposed/new VSP design intent check-in, but @emilywaggoner and other designers conducting those reviews may need some FE support to spot them. I don't think it would require an FE engineer at every review though; more just a deep knowledge of existing patterns and documentation about recommended (required?) components and patterns before anything is even mocked up. My assumption is that we don't do that kind of documentation currently. 
- I'm new here, and hopefully entering this conversation with humility and a desire to learn about what's already in place and why

**Comment by rianfowler** [comment]
@leahbannon
>Can y'all provide a couple plain language examples of the kinds of problems we're currently catching?
- VFS teams often don't know about the utilities available in the design system and add duplicate CSS to their apps. For example, teams will often create a new css class in their app to set the margin on a component when there are several utility classes [available](https://design.va.gov/utilities/margins) for setting the margin. This is why we will continue to require platform group approval for changes to any SCSS or CSS file on apps. 
- VFS sometimes need to be reminded to unit test their code. 
> Write up your hypotheses/KPIs first 
- KPIs
  - **website performance** (e.g. [first contentful paint, first input delay](https://developers.google.com/speed/pagespeed/insights/?url=www.va.gov)): we are currently [evaluating](https://github.com/department-of-veterans-affairs/va.gov-team/issues/4187) 3rd party tools to help us manage this. we're also figuring out how to change our processes to enforce good site performance e.g. checking bundle sizes during the CI, adding [synthetic monitoring](https://developer.mozilla.org/en-US/docs/Web/Performance/Rum-vs-Synthetic)
  - **code quality** - We've gotten feedback on the inconsistency in our code quality feedback. We plan on [evaluating](https://github.com/department-of-veterans-affairs/va.gov-team/issues/4091) third party code quality tools that can automatically detect issues that require manual review now (e.g. issues with code complexity). 
  - **unit test coverage** - we're using code climate to display this metric in PRs to teams but we aren't enforcing a coverage threshold. Efficacy of explicit unit test coverage % is debatable and we may get more mileage out of using a code quality tool. 
- hypotheses 
  - Without a high level of context, code reviews are a less effective way to enforce code quality and site performance. The scrutiny applied to reviews is inconsistent which adds uncertainty to the developer experience. The platform group often does not have useful feedback for application specific code due to lack of context. Removing required reviews from application specific code will improve the developer experience and allow the platform group to focus on tooling for enforcing code quality and performance metrics. 
> CSS review concerns / Size restrictions
- No plan for the automated check, yet. I think we would rather handle it further downstream in the performance monitoring. I think the goal should be enforcing [user-centric performance metrics](https://developers.google.com/web/fundamentals/performance/user-centric-performance-metrics). 
  - we also can't really enforce design consistency by ensuring they are using the design system utilities. We could prevent them from defining new colors but we couldn't ensure the layout looks "right" without seeing screenshots. 
> Automated code reviews 
- no timeline on the code quality tooling. I think the issue is more that we need to do some work on improving the consistency of our reviews (i.e. we need to get everyone on the same page), and that we think that time is better spent on evaluating third party tooling. In the mean time, we should pull back on doing the app specific reviews instead of subjecting the VFS teams to churn and uncertainty. 
> Team-level control
- I was going to defer to the VFS teams to do this. We don't have the GitHub team structure fleshed out but my understanding is that they can create sub groups in the write-access groups.

**Comment by rianfowler** [comment]
Tabling this for now.

**Pull request closed without merging**
